clang/lib/DPCT/ASTTraversal.cpp

```diff
@@ -8823,12 +8822,242 @@ void MemVarRule::registerMatcher(MatchFinder &MF) {
+          HasTypeCasted = true;
+        } else {
+          // Dim == 0
+          if (Decl->hasAttr<CUDAConstantAttr>() &&
+              (MemVarRef->getType()->getTypeClass() !=
+               Type::TypeClass::Elaborated)) {
+            const Expr *RHS = getRHSOfTheNonConstAssignedVar(MemVarRef);
+            if (RHS) {
+              auto Range = GetReplRange(RHS);
+              emplaceTransformation(new ReplaceText(
+                  Range.first, Range.second,
+                  buildString("const_cast<", RHS->getType(), ">(",
+                              ExprAnalysis::ref(RHS), ")")));
+              HasTypeCasted = true;
+            }
+          }
+        }
+      }
+    }
+    if (!HasTypeCasted && Decl->hasAttr<CUDAConstantAttr>() &&
+        (MemVarRef->getType()->getTypeClass() ==
+         Type::TypeClass::ConstantArray)) {
+      const Expr *RHS = getRHSOfTheNonConstAssignedVar(MemVarRef);
+      if (RHS) {
+        auto Range = GetReplRange(RHS);
+        emplaceTransformation(
+            new ReplaceText(Range.first, Range.second,
+                            buildString("const_cast<", RHS->getType(), ">(",
+                                        ExprAnalysis::ref(RHS), ")")));
+      }
+    }
+    auto VD = dyn_cast<VarDecl>(MemVarRef->getDecl());
+    if (Func->isImplicit() ||
+        Func->getTemplateSpecializationKind() == TSK_ImplicitInstantiation)
+      return;
+    if (VD == nullptr)
+      return;
+    auto Var = Global.findMemVarInfo(VD);
+    if (Func->hasAttr<CUDAGlobalAttr>() || Func->hasAttr<CUDADeviceAttr>()) {
+      if (DpctGlobalInfo::useGroupLocalMemory() &&
+          VD->hasAttr<CUDASharedAttr>() && VD->getStorageClass() != SC_Extern) {
+        if (!Var)
+          return;
+        if (auto B = dyn_cast_or_null<CompoundStmt>(Func->getBody())) {
+          if (B->body_empty())
+            return;
+          emplaceTransformation(new InsertBeforeStmt(
+              B->body_front(), Var->getDeclarationReplacement(VD)));
+          return;
+        }
+      }
+    } else {
+      if (Var && !VD->getType()->isArrayType() &&
+          VD->hasAttr<HIPManagedAttr>()) {
+        emplaceTransformation(new InsertAfterStmt(MemVarRef, "[0]"));
+      }
+    }
+  }
+}
 
+REGISTER_RULE(MemVarRefRule, PassKind::PK_Migration)
+
+void ConstantMemVarRule::registerMatcher(MatchFinder &MF) {
+  auto DeclMatcher =
+      varDecl(hasAttr(attr::CUDAConstant),
+              unless(hasAnyName("threadIdx", "blockDim", "blockIdx", "gridDim",
+                                "warpSize")));
+  MF.addMatcher(DeclMatcher.bind("var"), this);
   MF.addMatcher(varDecl(hasParent(translationUnitDecl())).bind("hostGlobalVar"),
                 this);
 }
 
-void MemVarRule::previousHCurrentD(const VarDecl *VD, tooling::Replacement &R) {
+void ConstantMemVarRule::runRule(const MatchFinder::MatchResult &Result) {
+  std::string CanonicalType;
+  auto IsVarUsedByRuntimeSymbolAPI = [&](const VarDecl *VarD) {
```

Comment:
Add design of high level decision make on how to migrate constant var as comment.